### PR TITLE
solve raises error on duplicate symbols

### PR DESCRIPTION
**Pull request by smichr**
#### Brief description of what is fixed or changed

Rather than deal in some unique way with a solve request that contains duplicate symbols, the presence of duplicates now raises an error. This seems like a good idea because there is no reason to include duplicates and it *might* indicate a typo of the person entering the request. `nonlinsolve` raises an error, `linsolve` does not (and produces a strange output):
```python
>>> linsolve([x-1],x)
{(1,)}
>>> linsolve([x-1],x,y)
{(1, y)}
>>> linsolve([x-1],x,x)
{(x + 1, x + 1)}
```

#### Other comments

An audit-via-test of solve output has been included and highlights some gray areas that are not resolved here (but should be). Namely
1. should providing equations as a list (even of one equation) change the output?
```python
>>> solve(x)
[0]
>>> solve(x,x)
[0]
>>> solve([x],x)
{x: 0}
>>> solve([x])
{x: 0}
```
2. which multi-equation output should come back as a single dictionary. Here are 3 possiblities

a. the purely linear system `solve((x + y, x - y - 2)) -> {x:1, y: -1}`

```python
>>> solve((x + y, x - y - 2))
{x: 1, y: -1}
>>> solve((x + y, x - y - 2),x,y)
{x: 1, y: -1}
```

b. the system that has a single solution solution but contains nonlinear terms

```python
>>> solve((x**2 - y, x - 2))  # must return symbols in some way
[{x: 2, y: 4}]
>>> solve((x**2 - y, x - 2),x,y)
[(2, 4)]
```

c. a system that is linear in monotonic generators but nonlinear in symbols `solve((exp(x)-2,y-x)) -> {x: log(2), y: log(2)}`
```python
>>> solve((exp(x)-2,y-x))
{x: log(2), y: log(2)}
>>> solve((exp(x)-2,y-x),x,y)
{x: log(2), y: log(2)}
```
#### Release Notes

* solvers
  * `solve` raises an error if duplicate symbols are provided
  * corrected a case where set=True would send back an empty list of symbols; code which uses `set=True` and only tests whether the result is empty will now fail: `s, sol = solve(..., set=True)` will have empty `sol` if there is no solution.

### CODE REVIEW

File: sympy/solvers/tests/test_solvers.py [hunk 1]
```diff
@@ -114,6 +114,446 @@ def test_guess_transcendental():
...
+        [(-2, y), (2, y)],
+        [{x: -2}, {x: 2}],
+        ([x, y], {(2, y), (-2, y)})]
+    assert [solve(e, y, x, **f) for f in flags] == \
+        [[(y, -2), (y, 2)],
+        [{x: -2}, {x: 2}],
+        ([y, x], {(y, -2), (y, 2)})]
+    assert [solve([e], y, x, **f) for f in flags] == [
+        [(y, -2), (y, 2)],
+        [{x: -2}, {x: 2}],
```
Comment (oscarbenjamin) [inline sympy/solvers/tests/test_solvers.py:220]: Why are these two different?

File: sympy/solvers/tests/test_solvers.py [hunk 2]
```diff
@@ -114,6 +114,446 @@ def test_guess_transcendental():
...
 
+def test_solve_io():
+    lx = x
+    nx = x**2 - 4
+    lxly = x - y + 1
+    lxny = x + y**2 - 1
+    nxly = y + x**2 - 1
+    nxny = y**2 + x**2 - 1
+    flags = {}, dict(dict=True), dict(set=True)
+    e = lx
```
Comment (oscarbenjamin) [inline sympy/solvers/tests/test_solvers.py:125]: This test is quite hard to follow. I think that redefining `e` like this makes it harder. Why not just use `lx` below?

Also I think it can probably be made more compact. At leat when dict or set is True the output should be more predictable so we can compress those cases together into something like:
```python
eqs_sol = [
    # eq, sol, solx, soly, solxy, solyx
    (x, [{x:0}], [{x:0}], [], [{x:0}], [{x:0}]),
    (x - y + 1, [{x:y-1}], [{x:y-1}], [{y:x+1}], [{x:y-1}], [{y:x+1}]),
    ...
]
for eq, sol, solx, soly, solxy, solyx in eqs_sol:
    for eqi in [eq, [eq], Eq(eq, 0), [Eq(eq, 0)]:
        assert solve(eq, dict=True) == sol
        assert solve(eq, x, dict=True) == solx
        assert solve(eq, y, dict=True) == soly
        ...
```
That way it's easier to look at `eqs_sol` to see how the different cases compare.

File: sympy/solvers/tests/test_solvers.py (hunk 1 above)
Comment (smichr) [reply sympy/solvers/tests/test_solvers.py:220]: It's hard to tell from context what you are referring to. Are you asking why tuples are used at L219 and dictionaries at L220? Or are you asking why specifying the order `y, x` in L218 gives different results from using a set at L221? The latter is because if the order is indeterminate, the output cannot be used unambiguously to know which values correspond with which variables unless the ordering routine is known (and cf #23828 for how arbitrary that can be).

Comment (oscarbenjamin) [reply sympy/solvers/tests/test_solvers.py:220]: I think I highlighted the wrong examples (it's quite confusing to see which example is being tested where in this function). We have:
```python
In [1]: solve([x**2 - 4], y, x)
Out[1]: [(y, -2), (y, 2)]

In [2]: solve([x**2 - 4], y, x, set=True)
Out[2]: ([x, y], {(-2, y), (2, y)})
```
With `set=True` here a symbol ordering `[y, x]` has been provided through the ordering of the unknowns in `*args` but the example comes back using a different ordering `[x, y]`.

Comment (smichr) [reply sympy/solvers/tests/test_solvers.py:220]: The set form is canonical regardless of how you pass the symbols, so `...x, y, set=True` = `...y,x,set=True` = `...{x,y}, set=True`. It is not hard, however, to keep the order if the order is determinate since, without adding a new flag `tuple=True` that is a way to get full ordered tuples back from solve.

Comment (smichr) [reply sympy/solvers/tests/test_solvers.py:220]: That can be handled in a separate PR. The purpose of this is to

1. capture the solve io fingerprint so subsequent mods will flag any differences and to identify issues that should be resolved (like the case here that you identified)
2. raise error on duplicate symbols
3. apply simplify per request on a case that was being missed
4. correct output for set=True on the corner case identified

File: sympy/solvers/tests/test_solvers.py (hunk 2 above)
Comment (smichr) [reply sympy/solvers/tests/test_solvers.py]: Please check new format.

File: sympy/solvers/tests/test_solvers.py [hunk 3]
```diff
@@ -114,6 +114,262 @@ def test_guess_transcendental():
...
+    assert [solve([*e], y, x, **f) for f in flags] ==       [lyx,   [d],    syx]    # y, x
+    assert [solve([*e], {y, x}, **f) for f in flags] ==     [[d],   [d],    sxy]    # {x, y}
+    assert [solve([*e], x, z, y, **f) for f in flags] ==    [txzy,  [d],    sxzy]   # x, z (missing), y
+    assert [solve([*e], {x, z, y}, **f) for f in flags] ==  [[d],   [d],    sxy]    # {x, y, z (missing)}
+    e = a*x + b - 2*x - y
+    assert [solve(e, a, b, **f) for f in flags] == [
+		{a: 2, b: y},[{a: 2, b: y}],([a, b], {(2, y)})]
+    assert [solve(e - b + b**2, a, b, **f) for f in flags] == [
+		[((-b**2 + 2*x + y)/x, b)],[{a: (-b**2 + 2*x + y)/x}],(
+		[a, b], {((-b**2 + 2*x + y)/x, b)})]
```
Comment (smichr) [inline sympy/solvers/tests/test_solvers.py:370]:
```suggestion
    assert [solve(e, a, b, **f) for f in flags] == [
        {a: 2, b: y},[{a: 2, b: y}],([a, b], {(2, y)})]
    assert [solve(e - b + b**2, a, b, **f) for f in flags] == [
        [((-b**2 + 2*x + y)/x, b)],[{a: (-b**2 + 2*x + y)/x}],(
        [a, b], {((-b**2 + 2*x + y)/x, b)})]
```

File: sympy/solvers/tests/test_solvers.py [hunk 4]
```diff
@@ -114,6 +114,262 @@ def test_guess_transcendental():
...
+    lody = [{y: -rx}, {y: rx}]
+    sx = ([x], {(-ry,), (ry,)})
+    sx1 = ([x], {(ry,), (-ry,)})
+    sxy = ([x, y], {(-ry, y), (ry, y)})
+    syxy = ([y, x], {(rx, x), (-rx, x)})
+    syxx = ([y, x], {(y, -ry), (y, ry)})
+    tm = (-ry, z, y)
+    tp = (ry, z, y)
+    vx = [-ry, ry]
+    tx = [(-ry,) (ry,)]
```
Comment (smichr) [inline sympy/solvers/tests/test_solvers.py:297]:
```suggestion
    tx = [(-ry,), (ry,)]
```

File: sympy/polys/numberfields/tests/test_minpoly.py [hunk 5]
```diff
@@ -419,10 +419,11 @@ def test_issue_18248():
...
     c1 = Circle(Point2D(3, sqrt(5)), 5)
     c2 = Circle(Point2D(4, sqrt(7)), 6)
-    assert intersection(c1, c2) == [Point2D(-1 + (-sqrt(7) + sqrt(5))*(-2*sqrt(7)/29
-    + 9*sqrt(5)/29 + sqrt(196*sqrt(35) + 1941)/29), -2*sqrt(7)/29 + 9*sqrt(5)/29
-    + sqrt(196*sqrt(35) + 1941)/29), Point2D(-1 + (-sqrt(7) + sqrt(5))*(-sqrt(196*sqrt(35)
-    + 1941)/29 - 2*sqrt(7)/29 + 9*sqrt(5)/29), -sqrt(196*sqrt(35) + 1941)/29 - 2*sqrt(7)/29 + 9*sqrt(5)/29)]
+    s = sqrt(196*sqrt(35) + 1941)/29
+    y = -2*sqrt(7)/29 + 9*sqrt(5)/29
+    r = (-sqrt(7) + sqrt(5))*(29*y + 29*s)/29
+    assert intersection(c1, c2) == [Point2D(-1 + r, y + s), Point2D(-1 - r, y - s)]
```
Comment (smichr) [inline sympy/polys/numberfields/tests/test_minpoly.py:425]:
```suggestion
    assert intersection(c1, c2) == [Point2D(-1 + r, y + s), Point2D(-1 - r, y - s)], intersection(c1, c2)
```

File: sympy/polys/numberfields/tests/test_minpoly.py [hunk 6]
```diff
@@ -419,10 +419,11 @@ def test_issue_18248():
...
     c1 = Circle(Point2D(3, sqrt(5)), 5)
     c2 = Circle(Point2D(4, sqrt(7)), 6)
-    assert intersection(c1, c2) == [Point2D(-1 + (-sqrt(7) + sqrt(5))*(-2*sqrt(7)/29
-    + 9*sqrt(5)/29 + sqrt(196*sqrt(35) + 1941)/29), -2*sqrt(7)/29 + 9*sqrt(5)/29
-    + sqrt(196*sqrt(35) + 1941)/29), Point2D(-1 + (-sqrt(7) + sqrt(5))*(-sqrt(196*sqrt(35)
-    + 1941)/29 - 2*sqrt(7)/29 + 9*sqrt(5)/29), -sqrt(196*sqrt(35) + 1941)/29 - 2*sqrt(7)/29 + 9*sqrt(5)/29)]
+    s = sqrt(196*sqrt(35) + 1941)/29
+    y = -2*sqrt(7)/29 + 9*sqrt(5)/29
+    r = (-sqrt(7) + sqrt(5))*(29*y + 29*s)/29
+    assert intersection(c1, c2) == [Point2D(-1 + r, y + s), Point2D(-1 - r, y - s)], intersection(c1, c2)
```
Comment (smichr) [inline sympy/polys/numberfields/tests/test_minpoly.py:425]:
```suggestion
    s = sqrt(196*sqrt(35) + 1941)
    d = (-sqrt(7) + sqrt(5))
    d2 = -2*sqrt(7)/29 + 9*sqrt(5)/29
    d3 = -9*sqrt(5) + 2*sqrt(7)
    assert intersection(c1, c2) == [Point2D(-1 - d*(s + d3)/29, d2 - s/29),
        Point2D(-1 + d*(s - d3)/29, d2 + s/29)]
```

**Comment by oscarbenjamin** [comment]
This PR seems to add many hours to the time taken to run the test suite.

**Comment by oscarbenjamin** [comment]
This has made `test_issue_6528` very slow.

File: sympy/solvers/solvers.py [hunk 7]
```diff
@@ -2051,8 +2071,12 @@ def key(sym):
...
                         result.remove(b)
 
-    default_simplify = bool(failed)  # rely on system-solvers to simplify
-    if  flags.get('simplify', default_simplify):
+    if not result:
+        return []
+
+    # rely on linear system solvers to simplify
+    default_simplify = bool(failed) or not linear
+    if flags.get('simplify', default_simplify):
```
Comment (oscarbenjamin) [inline sympy/solvers/solvers.py:2079]: This seems to be the problem. Why has `default_simplify` been changed?

File: sympy/solvers/tests/test_solvers.py [hunk 8]
```diff
@@ -114,6 +114,262 @@ def test_guess_transcendental():
...
+
+    d = {x: 0}
+    s = ([x], {(0,)})
+    syx = ([y, x], {(y, 0)})
+    sy = ([y], set())
+    vx = [0]
+    vyx = [(y, 0)]
+    e =                                                                         x
+    assert [solve(e, **f) for f in flags] ==            [vx,    [d],    s]      # no syms
+    assert [solve([e], **f) for f in flags] ==          [d,     [d],    s]
```
Comment (oscarbenjamin) [inline sympy/solvers/tests/test_solvers.py:132]: These tests are horrible. I gave feedback that they should be written in a way that was more readable. The number one goal in readability of tests should be to make it so that the input and output are close together so that they can be compared.

**Comment by oscarbenjamin** [comment]
@smichr please don't merge PRs without review.

**Comment by oscarbenjamin** [comment]
Also it is better if PRs don't mix orthogonal things together. From the title of this PR I would not have guessed what is actually changed in the diff. It looks like there are several bits here that should have been separate PRs.

I've opened #23888 to revert this PR.

**Comment by oscarbenjamin** [comment]
I've reverted this in #23888. The benchmark results there show that is gives a significant speed up relative to the changes here (https://github.com/sympy/sympy/pull/23888#issuecomment-1207523985). I'm not sure why that slowdown didn't show up in the benchmark results in this PR.

File: sympy/solvers/solvers.py (hunk 7 above)
Comment (smichr) [reply sympy/solvers/solvers.py:2079]: This was changed because the nonlinear solutions were not being simplified.
